Add a way to remove a single event listener

ypujante opened the pull request:
I know it has been quite a while since I opened #20983 but I finally got the time to work on it. I followed the discussion we had at the time and did the following:

- added `emscripten_remove_callback`
  - iterate over all handlers and only remove the one matching the arguments provided
- changed all event setters to pass `userData` and `eventTypeId` which are required by this API
- added a test that shows that it works

This first pass is to make sure that this would be an ok change to Emscripten and that the changes work (no failures introduced by the changes). I can obviously change whatever you want or drop it entirely if this is not desired. And can work on documentation/examples if that gets approval.

I believe it is a good addition because it allows to remove a previously set listener, which is not possible at the moment, as the APIs provided will remove all of the ones set (see discussion in the #20983 ticket).

Fixes: #20983

ypujante commented:
I am done with traveling so I can work on this again. And I would like to wrap this up if at all possible. 

@sbc100 do you have any opinion on @juj feedback? Do you have an opinion on removing the `userData` parameter?

@juj Can you provide feedback on my comment about pthread?

Thanks

sbc100 commented on src/lib/libhtml5.js:
```diff
@@ -412,6 +433,18 @@ var LibraryHTML5 = {
 
+  emscripten_remove_callback__proxy: 'sync',
+  emscripten_remove_callback__deps: ['$JSEvents', '$findEventTarget'],
+  emscripten_remove_callback: (target, userData, eventTypeId, callback) => {
```
I think this makes sense to me, but in that case I think it would also be good it if we make it impossible to register the same callback function  for the same event with different userData.  

```
emscripten_set_click_callback(window, data1, 0, mycallback);   // success
emscripten_set_click_callback(window, data2, 0, mycallback);   // error `mycallback` is already registered as a click callback
```

If we didn't error in the second case, and allowed the same callback to be registred twice, there would be no way to unregister just one of them.

ypujante replied:
* if we want to error when registering the same callback with a different user data, then it means we need to store the user data (which we don't prior to this changes)
* it also means that code that is currently working could fail after this change

I think what @juj suggested was to indeed un-register all callbacks that were registered with different data but with the same target/event_id/callback combination. That would have eliminated having to store the user data.

It sounds like you don't like this because then "there would be no way to unregister just one of them".

I am definitely not in favor of erroring when registering 2 identical callbacks with different user data, because from a developer point of view, this is absolutely not an error and I can imagine scenarios where that could be useful. 

So I guess the solution is to leave the API and changes as implemented in this PR. That way we can un-register a single callback (which IS the point of this PR). So I am personally in favor of this.

I will now work on updating the documentation and fixing the "Code Size" failure.

ypujante replied:
This a theoretical example, because I am not implementing it this way at the moment, but imagine for GLFW, you have 2 windows (represented in C by GLFWwindow pointers and canvases in HTML) and you need to set mouse move callbacks to track the mouse movements.

In order to detect moving the mouse outside the canvas while the button is clicked (something my library supports where others don't) you cannot set the callback on each canvas, it has to be set on the window... so the callbacks would be:

```cpp
emscripten_set_mousemove_callback(WINDOW, glfwWindow1, true, handle_move_callback);
emscripten_set_mousemove_callback(WINDOW, glfwWindow2, true, handle_move_callback);
```

Again I am not doing it this way, but that could be a way to implement it...

ypujante commented:
I think I am done on my side:

Based on the discussion, I left the code as-is with the `userData` parameter included. If we exclude it, then we need to disallow registering multiple callbacks with the same `userData`. It is still doable if you prefer to go down this route, but it means that it will break code relying on the fact that it is allowed now.

* I have changed the code
* Updated the documentation
* Updated the Changelog
* Fixed the size (after many attempts...)
* All tests are green

Let me know if there is anything else or if this can be merged as-is.

sbc100 commented:
Agree that adding 35 new API calls also seems like a bad idea :-/

Maybe we should at at least add a new generic API for this?    i.e. if we also add a single API for registering callbacks then we could at least say that the old API is deprecated?     On the other hand we do have a general policy of avoiding adding API that do too much multi-plexing since they prevent DCE.   Perhaps that is good argument for allowing an asymmetry in this case.

juj commented on site/source/docs/api_reference/html5.h.rst:
```diff
@@ -90,6 +90,50 @@ The ``useCapture`` parameter  maps to ``useCapture`` in `EventTarget.addEventLis
+Remove callback function
+------------------------
+
+In order to remove a callback, previously set via a ``emscripten_set_some_callback`` call, there is a dedicated and generic function for this purpose:
```
Maybe slightly simpler phrasing:

```
In order to unregister a single event handler callback, call the following function:
```

juj commented:
> Maybe we should at at least add a new generic API for this? i.e. if we also add a single API for registering callbacks then we could at least say that the old API is deprecated?

I think it would be good to avoid trying to eat too big pie here at once. @ypujante's use case is looking to be able to remove an individual callback, so it would be good to not snowball this into needing to redesign the whole API at once.

The name `emscripten_html5_remove_event_listener()` sounds good to me.

> > and also this would be good to support operation from pthreads to be symmetric/complete.
> 
> At first I tried to implement support for pthread but it doesn't seem to make sense from my understanding of the code. pthread is only used to specify how to invoke the callback once it has been registered. Registering the callback itself (which is done by calling `JSEvents.registerOrRemoveHandler(eventHandler);`) has nothing to do with pthread. So I don't see why we need pthread to remove/unregister the callback. Am I missing something?

Thanks for the note. I realize now that there is already a `emscripten_remove_callback__proxy: 'sync'` directive on the removal function handler, and that is indeed enough to make unregistering work from a pthread. Doesn't look like anything else is needed.

You had a good point at https://github.com/emscripten-core/emscripten/pull/25535/files#r2505004777 about why to keep the `userData` as part of the key to removing a callback. So it's a good idea to keep that.

So I would recommend:
- rename the function name to `emscripten_html5_remove_event_listener()`,
- keep `userData` as part of the key. Add a comment to the function docs to highlight this fact

That would work for me to land.

If we wanted to have a modern revision of the HTML5 event handler APIs that would have fewer registration functions, that can be a completely separate story from this one.

ypujante commented:
Thank you @juj for the review and comments. I am personally in full agreement with the approach. @sbc100 what do you think?

The only think I wanted to call out is the fact that this PR does not address the fact that a user can still call (for example) `emscripten_set_resize_callback` with `nullptr` for the function pointer to remove it. But as we know (that is the point of this PR), that it is wrong because it would remove **all** callbacks using the same target. I believe the solution is to not allow this anymore and instead use the new `emscripten_html5_remove_event_listener` function.

Obviously we could start by adding a deprecation warning when this scenario is detected. I could add this deprecation warning to this PR or open another one once this one is merged. Any preference?

sbc100 commented:
> Thank you @juj for the review and comments. I am personally in full agreement with the approach. @sbc100 what do you think?
> 

Fair enough.

> The only think I wanted to call out is the fact that this PR does not address the fact that a user can still call (for example) `emscripten_set_resize_callback` with `nullptr` for the function pointer to remove it. But as we know (that is the point of this PR), that it is wrong because it would remove **all** callbacks using the same target. I believe the solution is to not allow this anymore and instead use the new `emscripten_html5_remove_event_listener` function.
> 
> Obviously we could start by adding a deprecation warning when this scenario is detected. I could add this deprecation warning to this PR or open another one once this one is merged. Any preference?

Lets consider that (breaking) change separately, as a potentially followup.

ypujante commented:
@sbc100 I have implemented the changes discussed.

I have reverted the code size changes because it was polluting this PR with 30+ file changes and the codesize checks were still failing (a bit of a mystery to me). I can see that the failing codesize checks states that it is a warning that can be ignored, so I have chosen to ignore it...

There is only one test failing during package installation as I [reported earlier](https://github.com/emscripten-core/emscripten/pull/25799#issuecomment-3533012835) and it is clearly not related to these changes.

The relevant test is passing:

```
# from test-browser-firefox
  test_html5_remove_event_listener (test_browser.browser) ... ok (0.000s)
```

Let me know if there is anything else you want me to do for this PR

sbc100 commented on src/lib/libhtml5.js:
```diff
@@ -201,6 +201,25 @@ var LibraryHTML5 = {
+          && JSEvents.eventHandlers[i].eventTypeId === eventHandler.eventTypeId
+          && JSEvents.eventHandlers[i].callbackfunc === eventHandler.callbackfunc
+          && JSEvents.eventHandlers[i].userData === eventHandler.userData) {
+          JSEvents._removeHandler(i--);
```
Can you just `return {{{ cDefs.EMSCRIPTEN_RESULT_SUCCESS }}};` here?

And should we `return {{{ cDefs.EMSCRIPTEN_RESULT_INVALID_PARAM }}};`  below if its not found?

ypujante commented on test/test_html5_remove_event_listener.c:
```diff
@@ -0,0 +1,136 @@
+
+char const *userDataToString(void *userData) {
+  return userData ? (char const *) userData : "nullptr";
+}
```
Removed

sbc100 commented:
> > I can download the PR and upload the matching code size results for you.
> 
> That would be awesome. I wished I understood why it doesn't work on my machine as I am following the exact steps you are describing (including clearing the cache which I believe is very important). Does it matter if my regular install of emsdk is not the latest released one? (I cannot move to 4.0.18+ because you yanked support for `-sWEB_GPU` after only a few months of deprecating it and ImGui is currently not supporting it...)

As long as your emscripten checkout is configured to point the `tot` release of emsdk it should work.   I normally do this by setting `EM_CONFIG=/path/to/emsdk/.emscripten` before running clear cache + rebase.

sbc100 commented:
`source /usr/local/emsdk/emsdk_env.sh` should work just fine.

If you want to be able to just set `EM_CONFIG` then you need to make sure your python3 install has psutuil installed.   You can do that is serveral ways but one way is to do `pip4 install -r requirements-dev.txt`.    On my system its actually already installed as part of the `python3-psutil` package.

ypujante commented:
I am not sure I activated `tot`... (the instructions did not specify to activate it)... that might be my issue

juj commented:
The new added test fails on my CI boxes:

<img width="1525" height="505" alt="image" src="https://github.com/user-attachments/assets/1f816275-ee4f-48bf-ae5d-06c94fce2743" />

The failure occurs on Windows, Linux and macOS, but oddly only on "firefox latest beta" version of Firefox.

When I locally install that Firefox version (which is Firefox 144.0b6) via emsdk, and try out the failure, it never happens.

But for some reason on the CI, the failure is deterministic on all OSes.

The error reads `Assertion failed: count == eventHandlersCount`, so one of the registered event handler counts checks is failing.

I can't quite figure out right now what is different from local tests vs on the CI, except that the CI runs multiple tests in parallel (and with high contention), but locally, there is no contention, and it is just one test.

Curious if anyone would be able to reproduce this failure?

sbc100 commented:
I think the problem is `f'-DSAFARI_SUPPORT={int(not self.is_wasm64())}'` .. it doesn't take into account that you have extra browser settings in your `EMCC_CFLAGS` environment variable.

ypujante commented:
I assume if that is a problem for everybody, we can certainly revert https://github.com/emscripten-core/emscripten/pull/25858 which was just adding 1 test case for the 2 event handlers (and I assume this is what is causing problem with @juj setup, not the rest of the test)
